kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt

```diff
@@ -0,0 +1,887 @@
+  fun mutable(): KmValueParameter {
+    return KmValueParameter(flags, name).apply {
+      type = this@ImmutableKmValueParameter.type?.mutable()
+      varargElementType = this@ImmutableKmValueParameter.varargElementType?.mutable()
+    }
+  }
+}
+
+/** @return an immutable representation of this [KmTypeParameter]. */
+@KotlinPoetKm
+@JvmName("immutableOf")
+fun KmTypeParameter.immutable(): ImmutableKmTypeParameter {
+  return ImmutableKmTypeParameter(
+      flags,
+      name,
+      id,
+      variance,
+      upperBounds.map { it.immutable() },
+      annotations.unmodifiable()
+  )
+}
+
+/**
+ * Immutable representation of [KmTypeParameter].
+ *
+ * Represents a type parameter of a Kotlin class, function, property or type alias.
+ *
+ * @property flags type parameter flags, consisting of [Flag.TypeParameter] flags
+ * @property name the name of the type parameter
+ * @property id the id of the type parameter, useful to be able to uniquely identify the type parameter in different contexts where
+ *           the name isn't enough (e.g. `class A<T> { fun <T> foo(t: T) }`)
+ * @property variance the declaration-site variance of the type parameter
+ * @property upperBounds Upper bounds of the type parameter.
+ * @property annotations Annotations on the type parameter.
+ */
+@KotlinPoetKm
+data class ImmutableKmTypeParameter internal constructor(
+    override val flags: Flags,
+    val name: String,
+    val id: Int,
+    val variance: KmVariance,
+    val upperBounds: List<ImmutableKmType>,
+    val annotations: List<KmAnnotation>
+) : ImmutableKmWithFlags {
+  fun mutable(): KmTypeParameter {
+    return KmTypeParameter(flags, name, id, variance).apply {
+      upperBounds += this@ImmutableKmTypeParameter.upperBounds.map { it.mutable() }
+    }
+  }
+}
+
+/** @return an immutable representation of this [KmType]. */
+@KotlinPoetKm
+@JvmName("immutableOf")
+fun KmType.immutable(): ImmutableKmType {
+  return ImmutableKmType(
+      flags,
+      classifier,
+      arguments.map { it.immutable() },
+      abbreviatedType?.immutable(),
+      outerType?.immutable(),
+      flexibleTypeUpperBound?.immutable(),
+      isRaw,
+      annotations.unmodifiable()
+  )
+}
+
+/**
+ * Immutable representation of [KmType].
+ *
+ * Represents a type.
+ *
+ * @property flags type flags, consisting of [Flag.Type] flags
+ * @property classifier Classifier of the type.
+ * @property arguments Arguments of the type, if the type's classifier is a class or a type alias.
+ * @property isRaw `true` if the type is seen as a raw type in Java.
+ * @property annotations Annotations on the type.
+ */
+@KotlinPoetKm
+data class ImmutableKmType internal constructor(
+    override val flags: Flags,
+    val classifier: KmClassifier,
+    val arguments: List<ImmutableKmTypeProjection>,
+    /**
+     * Abbreviation of this type. Note that all types are expanded for metadata produced by the Kotlin compiler. For example:
+     *
+     *     typealias A<T> = MutableList<T>
```

Comment:
nit: enclose snippet in ```?

Reply:
It's markdown, so this is a code block with the indention. Was copied from the kotlinx-metadata doc